sycl/source/detail/kernel_program_cache.hpp

```diff
@@ -291,13 +359,194 @@ class KernelProgramCache {
 
   template <typename KeyT, typename ValT>
   void saveKernel(KeyT &&CacheKey, ValT &&CacheVal) {
+
+    if (SYCLConfig<SYCL_IN_MEM_CACHE_EVICTION_THRESHOLD>::
+            isProgramCacheEvictionEnabled()) {
+
+      ur_program_handle_t Program = std::get<3>(CacheVal);
+      // Save kernel in fast cache only if the corresponding program is also
+      // in the cache.
+      {
+        auto LockedCache = acquireCachedPrograms();
+        auto &ProgCache = LockedCache.get();
+        if (ProgCache.ProgramSizeMap.find(Program) ==
+            ProgCache.ProgramSizeMap.end())
+          return;
+      }
+
+      // Save reference between the program and the fast cache key.
+      MProgramToKernelFastCacheKeyMap[Program].emplace_back(CacheKey);
+    }
+
     std::unique_lock<std::mutex> Lock(MKernelFastCacheMutex);
     // if no insertion took place, thus some other thread has already inserted
     // smth in the cache
     traceKernel("Kernel inserted.", std::get<3>(CacheKey), true);
     MKernelFastCache.emplace(CacheKey, CacheVal);
   }
 
+  // Evict programs from cache to free up space.
+  void evictPrograms(size_t DesiredCacheSize, int CurrentCacheSize) {
+
+    // Figure out how many programs from the beginning we need to evict.
+    // [FIXME] Will the race on MCachedPrograms.Cache.empty() be a problem?
```

Comment:
I've restructured the code a little to avoid using lock here and still be functionally correct. After taking the `MProgramEvictionListMutex` lock, we check again the cache size (https://github.com/intel/llvm/pull/16062/files#diff-99114f8fd849025a6f86e930779900a3458758d8f5dd38e4a9bfbe21af7af015R407) and if it's smaller than the eviction threshold, we won't do any eviction.